### PR TITLE
Refactor and test `denitrification.py`

### LINKED ISSUE #488: refactor and test denitrification.py (soil/nitrogen_cycling) into SC_redesign directory

### PR DESCRIPTION
**Pull request by Isgotkowitz**
# Summary
This PR refactors and tests `denitrification.py`, which handles the denitrification of nitrates in the soil profile (denitrification is the process of bacteria turning NO3 (solid nitrates) in gases (N2 or N2O)), and is based on section 3:1.4 of SWAT.

# Main Changes
- `denitrification.py`
  - This module contains only two methods, one non-static main routine and one `@staticmethod`.
  - There is a TODO referencing issue #475 to revisit the method call associated with it, once the Carbon module has been fully sorted out.
  - Both methods are tested in `test_denitrification.py`
- `layer_data.py` 
  - 3 attributes have been added, one of them tracks the annual total of denitrified nitrates, and the others are used to determine how much denitrification occurs.
  - The annual total is reset by the `do_annual_reset()` method.

### CODE REVIEW

File: SC_redesign/Crop_and_Soil/soil/nitrogen_cycling/denitrification.py
Comment (morrowcj) [inline SC_redesign/Crop_and_Soil/soil/nitrogen_cycling/denitrification.py:52]: Has this been looked into yet?

File: SC_redesign/Crop_and_Soil/soil/nitrogen_cycling/denitrification.py
Comment (morrowcj) [inline SC_redesign/Crop_and_Soil/soil/nitrogen_cycling/denitrification.py:57]: Does this need to be bounded to 0? I'm genuinely asking.

```suggestion
            layer.nitrate_content -= max(nitrified_nitrates, 0)
```

If so, perhaps the bounding should be part of the return statement for `_calculate_nitrification_amount()` instead of here.

**Review by morrowcj — APPROVED** [review]
Looks good to me. Just a small comment about a `#TODO` and another about potentially preventing negative nitrification.

File: SC_redesign/Crop_and_Soil/soil/nitrogen_cycling/denitrification.py
Comment (Isgotkowitz) [reply SC_redesign/Crop_and_Soil/soil/nitrogen_cycling/denitrification.py:57]: I agree, I will make that change to `_calculate_nitrification_amount()`

File: SC_redesign/Crop_and_Soil/soil/nitrogen_cycling/denitrification.py
Comment (Isgotkowitz) [reply SC_redesign/Crop_and_Soil/soil/nitrogen_cycling/denitrification.py:52]: No, #475 is still unresolved. In this case it just depends on whether `soil_overall_carbon_fraction` represents only organic carbon or not.

File: SC_redesign/Crop_and_Soil/soil/layer_data.py
Comment (PooyaHekmati) [inline SC_redesign/Crop_and_Soil/soil/layer_data.py:261]: what happens if the value is not in this range? do have test coverage for such scenarios?

File: SC_redesign/Crop_and_Soil/soil/nitrogen_cycling/denitrification.py
Comment (PooyaHekmati) [inline SC_redesign/Crop_and_Soil/soil/nitrogen_cycling/denitrification.py]:
```suggestion
        """Initializes the SoilData object that this module will work with, or create one if none is provided.
```

Comment (PooyaHekmati) [reply SC_redesign/Crop_and_Soil/soil/nitrogen_cycling/denitrification.py]: Since soil_data can be passed, I assume it exists somewhere else as well. So, when we create it or otherwise manipulate it, how do we handle side effects?

File: SC_redesign/Crop_and_Soil/soil/nitrogen_cycling/denitrification.py
Comment (PooyaHekmati) [inline SC_redesign/Crop_and_Soil/soil/nitrogen_cycling/denitrification.py:26]: What if both are none?

File: SC_redesign/Crop_and_Soil/soil/nitrogen_cycling/denitrification.py
Comment (PooyaHekmati) [inline SC_redesign/Crop_and_Soil/soil/nitrogen_cycling/denitrification.py]: maybe something like this would help? Just an idea, feel free to use a better name or discard this comment 
```suggestion
            nutrient_is_below_threshold= layer.nutrient_cycling_water_factor < layer.denitrification_threshold_water_content:
            if nutrient_is_below_threshold:
                continue
```

File: SC_redesign/Crop_and_Soil/soil/nitrogen_cycling/denitrification.py
Comment (PooyaHekmati) [inline SC_redesign/Crop_and_Soil/soil/nitrogen_cycling/denitrification.py:57]: why?

File: SC_redesign/Crop_and_Soil/soil/nitrogen_cycling/denitrification.py
Comment (PooyaHekmati) [inline SC_redesign/Crop_and_Soil/soil/nitrogen_cycling/denitrification.py]: I don't have enough subject matter expertise to judge this name, maybe denitrification is a keyword. But if it is not, maybe 'denitrify' can be the name?

File: SC_redesign/Crop_and_Soil/soil/nitrogen_cycling/denitrification.py
Comment (PooyaHekmati) [inline SC_redesign/Crop_and_Soil/soil/nitrogen_cycling/denitrification.py:58]: same question: why? 
maybe the answer is obvious for an SME, but shedding more light on these two operations in the docstring might be a good idea.

File: SC_redesign/Crop_and_Soil/soil/nitrogen_cycling/denitrification.py
Comment (PooyaHekmati) [inline SC_redesign/Crop_and_Soil/soil/nitrogen_cycling/denitrification.py]: I don't think this comment is needed.

File: SC_redesign/Crop_and_Soil/soil/nitrogen_cycling/denitrification.py
Comment (PooyaHekmati) [inline SC_redesign/Crop_and_Soil/soil/nitrogen_cycling/denitrification.py]: would it make sense to bring this here?
```suggestion
        """Calculates the fraction of nitrates lost to nitrification as nitrification factor, and bounds it to be in the range [0.0, 1.0].
```
maybe explain how bounding works? do we clamp or normalize?

File: SC_redesign/Crop_and_Soil/soil/nitrogen_cycling/denitrification.py
Comment (PooyaHekmati) [inline SC_redesign/Crop_and_Soil/soil/nitrogen_cycling/denitrification.py]: maybe?
```suggestion
                                        temp_factor: float, organic_carbon_content_percent: float) -> float:
```

File: SC_redesign/Crop_and_Soil/soil/nitrogen_cycling/denitrification.py
Comment (PooyaHekmati) [inline SC_redesign/Crop_and_Soil/soil/nitrogen_cycling/denitrification.py]:
```suggestion
        References
        ----------
        SWAT Theoretical documentation eqn. 3:1.4.1

        """
```

Comment (PooyaHekmati) [inline SC_redesign/Crop_and_Soil/soil/nitrogen_cycling/denitrification.py]:
```suggestion
        References
        ----------
        SWAT Theoretical documentation eqn. 3:1.4.1

        """
```

File: SC_redesign/Crop_and_Soil/soil/nitrogen_cycling/denitrification.py
Comment (PooyaHekmati) [inline SC_redesign/Crop_and_Soil/soil/nitrogen_cycling/denitrification.py]:
```suggestion
        exponential_term = exp(-1 * denitrification_rate_coefficient * temp_factor * percent_organic_carbon_content)
        nitrification_factor = 1 - exponential_term
        clamped_nitrification_factor = max(min(1.0, (nitrification_factor )), 0.0)
        return nitrate_content * clamped_nitrification_factor 
```

File: SC_redesign/Crop_and_Soil/tests_SC/soil_tests/nitrogen_cycling_tests/test_denitrification.py
Comment (PooyaHekmati) [inline SC_redesign/Crop_and_Soil/tests_SC/soil_tests/nitrogen_cycling_tests/test_denitrification.py]: unnecessary

File: SC_redesign/Crop_and_Soil/tests_SC/soil_tests/nitrogen_cycling_tests/test_denitrification.py
Comment (PooyaHekmati) [inline SC_redesign/Crop_and_Soil/tests_SC/soil_tests/nitrogen_cycling_tests/test_denitrification.py]: we usually use 'actual' as the name of this variable.

File: SC_redesign/Crop_and_Soil/tests_SC/soil_tests/nitrogen_cycling_tests/test_denitrification.py
Comment (PooyaHekmati) [inline SC_redesign/Crop_and_Soil/tests_SC/soil_tests/nitrogen_cycling_tests/test_denitrification.py]: it would be better if you hardcode the expected values in the data provider and remove this calculation. 
Generally speaking, it is a good practice to avoid duplicating code logic in the test. Also, hardcoding the expected value is easier to read and judge if things are making sense.

File: SC_redesign/Crop_and_Soil/tests_SC/soil_tests/nitrogen_cycling_tests/test_denitrification.py
Comment (PooyaHekmati) [inline SC_redesign/Crop_and_Soil/tests_SC/soil_tests/nitrogen_cycling_tests/test_denitrification.py]: unnecessary

File: SC_redesign/Crop_and_Soil/tests_SC/soil_tests/nitrogen_cycling_tests/test_denitrification.py
Comment (PooyaHekmati) [inline SC_redesign/Crop_and_Soil/tests_SC/soil_tests/nitrogen_cycling_tests/test_denitrification.py]: there are lots of magic values in this test case, is it possible to explain them a bit? So, if I changed the code and this test failed, how do I know which value is doing what? How do I know if it is ok to change these values to make the test pass or not? What is the significance of indices 0, 2, 3? Why index 1 is handled differently?

**Review by PooyaHekmati — APPROVED** [review]
Approving to unblock the PR.
The PR is good enough to be merged, but I think there is room for improvement before getting merged into the master.

**Comment by PooyaHekmati** [comment]
For record, it took me an hour to review this.

File: SC_redesign/Crop_and_Soil/soil/layer_data.py
Comment (Isgotkowitz) [reply SC_redesign/Crop_and_Soil/soil/layer_data.py:261]: As of right now, I think it is ok to leave the `denitrification_rate_coefficient` invalidated because it is such an  advanced attribute that anyone who messes with it will probably be looking at what it is and what acceptable values for it are.

More generally, input validation is light in the current state of the Soil and Crop module, and I think it would be good to decide on a strategy for input validation / error raising before diving into it and/or adding it for each piece of data that comes up. Some considerations for said strategy:

1. It needs to be consistent with how errors are handled throughout the rest of RuFaS, and as of this writing I do not know that team standard.
2. If errors are going to be raised, where will they be propagated up to, if propagated up at all? If propagated up, there should be appropriate handling in the methods/modules etc.
3. What are the metrics for deciding whether are error is unrecoverable or not?
  - For example, with the `denitrification_rate_coefficient`, if a user tries to set it to 3.5, should an error be raised that they tried to set it to an invalid value and the simulation stopped, or set it to 3.0 (the closest valid value), send a warning to the user, and run the simulation?
4. For attributes/elements that potentially change everyday, should these values be checked in every method they get passed to, or would it be better to have a method that runs once a day to validate every value?

File: SC_redesign/Crop_and_Soil/soil/nitrogen_cycling/denitrification.py
Comment (Isgotkowitz) [reply SC_redesign/Crop_and_Soil/soil/nitrogen_cycling/denitrification.py]: I am not sure exactly what you mean by side effects. If a pre-existing `SoilData` object is not passed to this module and it creates its own, the module will run in isolation unless that newly-created `SoilData` object is extracted from this object and used in all other modules being run.

File: SC_redesign/Crop_and_Soil/soil/nitrogen_cycling/denitrification.py
Comment (Isgotkowitz) [reply SC_redesign/Crop_and_Soil/soil/nitrogen_cycling/denitrification.py:26]: It will first see that `soil_data` is None, then move to trying to create a new `SoilData` instance. When it does this, the `post_init` method of `SoilData` will see that it is none and throw an error.

File: SC_redesign/Crop_and_Soil/soil/nitrogen_cycling/denitrification.py
Comment (Isgotkowitz) [reply SC_redesign/Crop_and_Soil/soil/nitrogen_cycling/denitrification.py]: Good suggestion

File: SC_redesign/Crop_and_Soil/soil/nitrogen_cycling/denitrification.py
Comment (Isgotkowitz) [reply SC_redesign/Crop_and_Soil/soil/nitrogen_cycling/denitrification.py]: I am in the same boat in terms of subject matter expertise, but I think denitrify is accurate for the operations being conducted here and is a much better name that the current one.

File: SC_redesign/Crop_and_Soil/soil/nitrogen_cycling/denitrification.py
Comment (Isgotkowitz) [reply SC_redesign/Crop_and_Soil/soil/nitrogen_cycling/denitrification.py:58]: Good call

File: SC_redesign/Crop_and_Soil/soil/nitrogen_cycling/denitrification.py
Comment (Isgotkowitz) [reply SC_redesign/Crop_and_Soil/soil/nitrogen_cycling/denitrification.py]: It is a good suggestion, but I am going to keep it as `percent_organic_carbon_content` so that it stays in sync with the attribute names in `LayerData`. This can be addressed by #404 though.

File: SC_redesign/Crop_and_Soil/soil/nitrogen_cycling/denitrification.py
Comment (Isgotkowitz) [reply SC_redesign/Crop_and_Soil/soil/nitrogen_cycling/denitrification.py]: I like keeping comments about specific implementation details in the `Notes` section, IMO it keeps the docstring a bit cleaner. Also this method calculates the amount of nitrate lost, calculating the nitrification factor is one of the components of this process.

I am not sure what the difference between normalizing and clamping is (but if I had to guess, I'd say clamp). As for an explanation of why the bounds are what they are, I think it would be pretty obvious to a SME but adding a note wouldn't hurt.

File: SC_redesign/Crop_and_Soil/soil/nitrogen_cycling/denitrification.py
Comment (Isgotkowitz) [reply SC_redesign/Crop_and_Soil/soil/nitrogen_cycling/denitrification.py]: Good suggestion

File: SC_redesign/Crop_and_Soil/tests_SC/soil_tests/nitrogen_cycling_tests/test_denitrification.py
Comment (Isgotkowitz) [reply SC_redesign/Crop_and_Soil/tests_SC/soil_tests/nitrogen_cycling_tests/test_denitrification.py]: Good to know

File: SC_redesign/Crop_and_Soil/tests_SC/soil_tests/nitrogen_cycling_tests/test_denitrification.py
Comment (Isgotkowitz) [reply SC_redesign/Crop_and_Soil/tests_SC/soil_tests/nitrogen_cycling_tests/test_denitrification.py]: Interesting, I will use this testing strategy moving forward.

I am curious if there is a best practice for calculating those expected values, especially for a method that performs complex calculations? If I were going to determine expected values for this function outside of this test, I think I would write a method in interactive Python mode and use that to calculate the expected values. However this would still be problematic, because I still might make the same mistakes in both method implementations. Worse yet, if the method giving me the expected values was wrong and I don't have it saved then there wouldn't be a way for me to go back and check it as a possible reason for why my tests are failing.

File: SC_redesign/Crop_and_Soil/tests_SC/soil_tests/nitrogen_cycling_tests/test_denitrification.py
Comment (Isgotkowitz) [reply SC_redesign/Crop_and_Soil/tests_SC/soil_tests/nitrogen_cycling_tests/test_denitrification.py]: Good q's, I will add to the docstring about how this test achieves full coverage, which will explain those values.

**Pull request merged**
